### PR TITLE
Adding a helper method to validate versions in settings files

### PR DESCRIPTION
**Pull request by john-science**
## Description

There was [a request](https://github.com/terrapower/armi/issues/1095) to validate the versions of the app or plugin from a custom setting.

This is a little tool the user can use as part of a setting validation, to validate their versions for controlled runs.  Please see [this armi-example-app PR](https://github.com/terrapower/armi-example-app/pull/3) for how to do that.  

After both of these PRs are merged, I could also add examples of custom settings to the ARMI docs, under tutorials.

### Docs

I have also added a basic introduction to how to add new `Settings` to you `Plugin` in the docs.

---

## Checklist

- [X] This PR has only one purpose or idea.
- [X] Tests have been added/updated to verify that the new/changed code works.

- [X] The code style follows [good practices](https://terrapower.github.io/armi/developer/standards_and_practices.html).
- [X] The commit message(s) follow [good practices](https://terrapower.github.io/armi/developer/tooling.html).

- [X] The [release notes](https://terrapower.github.io/armi/release/index.html) (location `doc/release/0.X.rst`) are up-to-date with any bug fixes or new features.
- [X] The documentation is still up-to-date in the `doc` folder.
- [X] The dependencies are still up-to-date in `setup.py`.

### CODE REVIEW

**Comment by john-science** [comment]
@drewj-usnctech This is the PR for our little helper function.

Sorry, THIS we can do today. But it looks like it would take a medium-size overhaul of the settings system to write directly to the `metadata` header in the settings file.

We can talk about that, if it is high value to you.

File: armi/operators/settingsValidation.py [hunk 1]
```diff
@@ -726,3 +727,45 @@ def createQueryRevertBadPathToDefault(inspector, settingName, initialLambda=None
...
+    elif re.search(fullV, versionRequired) is not None:
+        return versionThis == versionRequired
+    elif re.search(medV, versionRequired) is not None:
+        return versionThis[:-2] == versionRequired
+    elif re.search(minV, versionRequired) is not None:
+        return versionThis[0] == versionRequired
+    else:
+        raise ValueError(
+            "The required version is not a valid format: {}".format(versionRequired)
+        )
```
Comment (keckler) [inline armi/operators/settingsValidation.py:771]: Is this ValueError duplicating the one above?

File: armi/operators/settingsValidation.py [hunk 2]
```diff
@@ -726,3 +727,45 @@ def createQueryRevertBadPathToDefault(inspector, settingName, initialLambda=None
...
+        This must be in one of the forms: 1.2.3, 1.2, or 1
+
+    Returns
+    -------
+    bool
+        Does this version match the version in the Settings file/object?
+    """
+    fullV = "[0-9]\.[0-9]\.[0-9]"
+    medV = "[0-9]\.[0-9]"
+    minV = "[0-9]"
```
Comment (keckler) [inline armi/operators/settingsValidation.py:751]: I think these patterns need to be extended to handle numbers with multiple digits?

File: armi/operators/settingsValidation.py (hunk 1 above)
Comment (john-science) [reply armi/operators/settingsValidation.py:771]: There are two `ValueError`s here. One for `versionThis` not being valid, and one for `versionRequired` not being valid.

File: armi/operators/settingsValidation.py (hunk 2 above)
Comment (john-science) [reply armi/operators/settingsValidation.py:751]: AH crap.

Yes.  You're right. Dammit. Thanks!

Comment (john-science) [reply armi/operators/settingsValidation.py]: Fixed!

**Comment by john-science** [comment]
@drewj-usnctech   I know it doesn't get us 100% the way to putting your version in `metadata`, but here is the PR.  Frankly, I think it would be easier to remove the `metadata` section from the `Settings` files entirely than to fully support them, so we could add more settings there.

Tell me what you think. Particularly, does my addition to the tutorial docs make enough sense?

**Review by opotowsky — APPROVED** [review]
looks good!

File: doc/tutorials/making_your_first_app.rst [hunk 3]
```diff
@@ -164,6 +164,27 @@ add a little more to the plugin.
...
+runs.
+
+The following example boiler plate code defines three settings. We define two simple
+number settings (inlet and outlet temperatures), and we use :py:class:`Query 
+<armi.operators.settingsValidation .Query>` to define validation on those settings. Here,
+the validation isn't very exciting, we just make sure the temperatures are above zero.
+That's not particularly physically meaningful, but serves as a simple example. The next
+setting is a little more complicated, we define a setting ``myAppVersion`` that defines
+a specific version of our app that this setting file is valid for. And if you try to run
+a different version you get a nasty warning printed to the screen.
```
Comment (drewj-usnctech) [inline doc/tutorials/making_your_first_app.rst:181]: Just so I'm clear, it's now on the plugin / app side to raise an error if they so choose if versions are unsupported?

- Plugin provides a setting for some version
- Plugin defines a setting validator

This works, just want to make sure I understand the paradigm

Comment (john-science) [reply doc/tutorials/making_your_first_app.rst:181]: Right now all the `settingsValidation` does is give the option for the user to hand-change a setting interactively. They don't kill a run.  e.g.:

https://github.com/terrapower/armi/blob/e0b01d708eac7ba73c0caabde75158af97fa007e/armi/operators/settingsValidation.py#L51-L61

So, that is the default workflow that I want to show in the `armi-example-app`.

But, in YOUR case, I imagine you would do in the `condition` step of your `Query` is `raise ValueError()` before you `return False`.  That way your run would _actually_ stop if the version was wrong.  Remember, you can put arbitrary code into that first `"condition"` field in a `Query`.

Sorry, I was trying to tow the line between what a general user of the `armi-example-app` would like, and what I thought would be useful to you.

Does that make sense?  Does that seem like a pathway that gets you everything you need/want?

Comment (drewj-usnctech) [reply doc/tutorials/making_your_first_app.rst:181]: Yeah that makes sense and should work. Thanks!

I think also if you use `correction=inspector.NO_ACTION` the `Query` validation will error for you

Comment (john-science) [reply doc/tutorials/making_your_first_app.rst:181]: Oh, possibly you are right. I should try that.

How do you feel about this PR?

I know it's not 100% what you wanted, but will it be helpful?  Should we merge it?

Comment (drewj-usnctech) [reply doc/tutorials/making_your_first_app.rst:181]: Oh yes, this is very helpful! The background, context, and implementation all makes a lot more sense.

I'm open to discussing how to improve this later, but I'm also more than content with where this is at

**Pull request merged**